app/src/src_nonfreenet/org/breezyweather/sources/bmkg/BmkgService.kt

```diff
@@ -0,0 +1,286 @@
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, version 3 of the License.
+ *
+ * Breezy Weather is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package org.breezyweather.sources.bmkg
+
+import android.content.Context
+import android.graphics.Color
+import breezyweather.domain.location.model.Location
+import breezyweather.domain.weather.wrappers.SecondaryWeatherWrapper
+import breezyweather.domain.weather.wrappers.WeatherWrapper
+import dagger.hilt.android.qualifiers.ApplicationContext
+import io.reactivex.rxjava3.core.Observable
+import org.breezyweather.common.exceptions.SecondaryWeatherException
+import org.breezyweather.common.source.HttpSource
+import org.breezyweather.common.source.MainWeatherSource
+import org.breezyweather.common.source.ReverseGeocodingSource
+import org.breezyweather.common.source.SecondaryWeatherSource
+import org.breezyweather.common.source.SecondaryWeatherSourceFeature
+import org.breezyweather.sources.bmkg.json.BmkgCurrentResult
+import org.breezyweather.sources.bmkg.json.BmkgForecastResult
+import org.breezyweather.sources.bmkg.json.BmkgIbfResult
+import org.breezyweather.sources.bmkg.json.BmkgWarningResult
+import retrofit2.Retrofit
+import javax.inject.Inject
+import javax.inject.Named
+
+class BmkgService @Inject constructor(
+    @ApplicationContext context: Context,
+    @Named("JsonClient") client: Retrofit.Builder
+) : HttpSource(), MainWeatherSource, SecondaryWeatherSource, ReverseGeocodingSource {
+
+    override val id = "bmkg"
+    override val name = "BMKG"
+    override val privacyPolicyUrl = ""
+
+    override val color = Color.rgb(0, 153, 0)
+    override val weatherAttribution = "Badan Meteorologi, Klimatologi, dan Geofisika"
+
+    private val mApi by lazy {
+        client
+            .baseUrl(BMKG_BASE_URL)
+            .build()
+            .create(BmkgApi::class.java)
+    }
+
+    override val supportedFeaturesInMain = listOf(
+        SecondaryWeatherSourceFeature.FEATURE_CURRENT,
+        SecondaryWeatherSourceFeature.FEATURE_ALERT
+    )
+
+    override fun isFeatureSupportedInMainForLocation(
+        location: Location,
+        feature: SecondaryWeatherSourceFeature?
+    ): Boolean {
+        return location.countryCode.equals("ID", ignoreCase = true)
+    }
+
+    override fun requestWeather(
+        context: Context,
+        location: Location,
+        ignoreFeatures: List<SecondaryWeatherSourceFeature>
+    ): Observable<WeatherWrapper> {
+        // API Key is needed for warnings, but not for current/forecast.
+        val apiKey = BMKG_API_KEY
```

Comment:
Make it a preference so that the user can change it if it suddenly stops working

Reply:
Fixed.